### ISSUE TITLE
Logging Source Generator fails to compile due to CS8795 error with file-scoped namespaces

### ISSUE THREAD

**Issue by martincostello** [dotnet/runtime#57880]
### Description

If a class used to provide the partial methods for the Logging Source Generator uses file-scoped namespaces, the project fails to compile with errors similar to the below.

```
Project\Logger.cs(11,36): error CS8795: Partial method 'Logger.Greeting(ILogger, string)' must have an implementation part because it has accessibility modifiers. [Project\Project.csproj]
Project\Microsoft.Extensions.Logging.Generators\Microsoft.Extensions.Logging.Generators.LoggerMessageGenerator\LoggerMessage.g.cs(11,36): error CS0759: No defining declaration found for implementing declaration of partial method 'Logger.Greeting(ILogger, string)' [Project\Project.csproj]
```

#### `Project.csproj`

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net6.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="6.0.0-rc.1.21420.7" />
    <PackageReference Include="Microsoft.Extensions.Logging" Version="6.0.0-rc.1.21420.7" />
    <PackageReference Include="Microsoft.Extensions.Logging.Console" Version="6.0.0-rc.1.21420.7" />
  </ItemGroup>
</Project>
```

#### `Program.cs`

```c#
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;
using MyLibrary;

using var serviceProvider = new ServiceCollection()
    .AddLogging(builder => builder.AddConsole())
    .BuildServiceProvider();

var loggerFactory = serviceProvider.GetRequiredService<ILoggerFactory>();
var logger = loggerFactory.CreateLogger("MyLogger");

Logger.Greeting(logger, "World");
```

#### `Logger.cs`

```c#
using Microsoft.Extensions.Logging;

namespace MyLibrary;

internal partial class Logger
{
    [LoggerMessage(EventId = 1, Level = LogLevel.Information, Message = "Hello {Name}!")]
    public static partial void Greeting(ILogger logger, string name);
}
```

#### Expected Output

```sh
Project> dotnet run
info: MyLogger[1]
      Hello World!
```

### Configuration

.NET SDK `6.0.100-rc.1.21420.39`
Microsoft.Extensions.Logging `6.0.0-rc.1.21420.7`

### Regression?

No.

**Comment by ylabade** [comment]
Potential fix in `LoggerMessageGenerator.Parser.cs` line 410 :
```diff
// determine the namespace the class is declared in, if any
SyntaxNode? potentialNamespaceParent = classDec.Parent;
- while (potentialNamespaceParent != null && potentialNamespaceParent is not NamespaceDeclarationSyntax)
+ while (potentialNamespaceParent != null &&
+    potentialNamespaceParent is not NamespaceDeclarationSyntax &&
+    potentialNamespaceParent is not FileScopedNamespaceDeclarationSyntax)
{
    potentialNamespaceParent = potentialNamespaceParent.Parent;
}
- if (potentialNamespaceParent is NamespaceDeclarationSyntax namespaceParent)
+ if (potentialNamespaceParent is BaseNamespaceDeclarationSyntax namespaceParent)
{
```

The type `FileScopedNamespaceDeclarationSyntax` seems to be available from `Microsoft.CodeAnalysis.CSharp` version 4.0.0-3.final

**Comment by davidfowl** [comment]
Seems like we should fix this for 6.0.0

cc @maryamariyan @ericstj @eerhardt

**Comment by martincostello** [comment]
I've opened a PR with the proposed fix https://github.com/dotnet/runtime/issues/57880#issuecomment-903192529 in #57894.

**Comment by maryamariyan** [comment]
Submitted the fix to main, will need to backport to 6.0.0 and test/fix for JSON if needed too. 

/cc @layomia

**Comment by martincostello** [comment]
JSON seems fine: https://github.com/dotnet/runtime/pull/57894#issuecomment-917380776

**Comment by tarekgh** [comment]
Has been fixed by #57894
